zoidberg/weights.py

```diff
@@ -0,0 +1,221 @@
+    for i in range(MXG):
+        for j in range(ny):
+            for k in range(nz):
+                cell_number_array[i, j, k] = cell_number
+                cell_number += 1
+    # Outer radial boundary cells
+    for i in range(nx - MXG, nx):
+        for j in range(ny):
+            for k in range(nz):
+                cell_number_array[i, j, k] = cell_number
+                cell_number += 1
+
+    backward_cell_number = np.zeros((nx, ny, nz), dtype=int)
+    forward_cell_number = np.zeros((nx, ny, nz), dtype=int)
+
+    # Iterate through for forward and backward maps
+    forward_xt_prime = maps["forward_xt_prime"]
+    backward_xt_prime = maps["backward_xt_prime"]
+
+    # Number of radial boundary cells
+    N_radial = 2 * MXG * ny * nz
+    cell_number = N_evolving + N_radial
+
+    # Add the parallel boundary cells
+    for i in range(MXG, nx - MXG):
+        for j in range(ny):
+            for k in range(nz):
+                if backward_xt_prime[i, j, k] < 0.0:
+                    backward_cell_number[i, j, k] = cell_number
+                    cell_number += 1
+                elif backward_xt_prime[i, j, k] >= nx:
+                    backward_cell_number[i, j, k] = cell_number
+                    cell_number += 1
+                if forward_xt_prime[i, j, k] < 0.0:
+                    forward_cell_number[i, j, k] = cell_number
+                    cell_number += 1
+                elif forward_xt_prime[i, j, k] >= nx:
+                    forward_cell_number[i, j, k] = cell_number
+                    cell_number += 1
+
+    return {
+        "N_cells": cell_number,
+        "N_evolving": N_evolving,
+        "cell_number": cell_number_array,
+        "forward_cell_number": forward_cell_number,
+        "backward_cell_number": backward_cell_number,
+    }
+
+
+def calc_interpolation(cell_number, MXG, yoffset, xtarr, ztarr):
+    """
+    Calculate CSR format matrix representing a 2D (X-Z) interpolation operation
+
+    Implements the cubic Catmull-Rom spline
+    Coefficients taken from https://en.wikipedia.org/wiki/Cubic_Hermite_spline
+    """
+
+    # Offsets and weights for 1D interpolation
+    offsets1D = [-1, 0, 1, 2]
+
+    def weights1D(u):
+        return 0.5 * np.array(
+            [
+                -(u**3) + 2.0 * u**2 - u,
+                3.0 * u**3 - 5.0 * u**2 + 2,
+                -3.0 * u**3 + 4.0 * u**2 + u,
+                u**3 - u**2,
+            ]
+        )
+
+    # CSR format
+    weights = []
+    columns = []
+    rows = []
+
+    nx, ny, nz = cell_number.shape
+
+    weight_number = 0  # Track location in weights & columns arrays
+    for i in range(MXG, nx - MXG):
+        for j in range(ny):
+            for k in range(nz):
+                xt = xtarr[i, j, k]
+                zt = ztarr[i, j, k]
+                if (xt < 0.0) or (xt >= nx):
+                    # Boundary
+                    rows.append(-1)
+                else:
+                    # Not a boundary point => Interpolating
```

Comment:
```suggestion
                    rows.append(-1)
                    continue
                # Not a boundary point => Interpolating
```
you can avoid some indention with `continue`